lib/widgets/swipable_message_row.dart

```diff
@@ -0,0 +1,115 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter_gen/gen_l10n/zulip_localizations.dart';
+
+import '../api/model/model.dart';
+import 'icons.dart';
+import 'text.dart';
+import 'theme.dart';
+
+class SwipableMessageRow extends StatefulWidget {
+  const SwipableMessageRow({
+    super.key,
+    required this.children,
+    required this.message,
+  });
+
+  final List<Widget> children;
+  final Message message;
+
+  @override
+  State<StatefulWidget> createState() => _SwipableMessageRowState();
+}
+
+class _SwipableMessageRowState extends State<SwipableMessageRow> {
+  @override
+  Widget build(BuildContext context) {
+    final hasMarker = widget.message.editState != MessageEditState.none;
+
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.baseline,
+      textBaseline: localizedTextBaseline(context),
+      children: [
+        hasMarker
+          ? _EditStateMarker(editState: widget.message.editState)
+          : const SizedBox(width: 16),
+        ...widget.children,
+      ],
+    );
+  }
+}
+
+class _EditStateMarker extends StatelessWidget {
+  /// The minimum width of the marker.
+  ///
+  /// Currently, only the collapsed state of the marker has been implemented,
+  /// where only the marker icon, not the marker text, is visible.
+  static const double widthCollapsed = 16;
+
+  const _EditStateMarker({
+    required MessageEditState editState,
+  }) : _editState = editState;
+
+  final MessageEditState _editState;
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = DesignVariables.of(context);
+    final zulipLocalizations = ZulipLocalizations.of(context);
+
+    final IconData icon;
+    final double iconSize;
+    final String markerText;
+
+    switch (_editState) {
+      case MessageEditState.none:
+        return const SizedBox(width: widthCollapsed);
+      case MessageEditState.edited:
+        icon = ZulipIcons.edited;
+        iconSize = 14;
+        markerText = zulipLocalizations.messageIsEdited;
+        break;
+      case MessageEditState.moved:
+        icon = ZulipIcons.message_moved;
+        iconSize = 8;
+        markerText = zulipLocalizations.messageIsMoved;
+        break;
+    }
+
+    var marker = Row(
+      mainAxisAlignment: MainAxisAlignment.end,
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        Flexible(
+          fit: FlexFit.loose,
+          // For now, [markerText] is not displayed because it is transparent and
+          // there is not enough space in the parent ConstrainedBox.
+          child: Text('$markerText ',
```

Comment:
Let's leave this out of this static version, since it's not yet being used — that will make the logic easier to think about. Then we can introduce it (modulo other edits, like Chris's comment about the space character vs. an explicit number of pixels of spacing) in the followup PR #765 that adds the swipe-to-reveal feature.